src/main/java/picard/nio/PicardBucketUtils.java

```diff
@@ -67,19 +65,63 @@ public static PicardHtsPath getTempFilePath(final String directory, String prefi
         }
     }
 
-    // For local temp file, directory should be null.
+    /**
+     * This overload of getTempFilePath takes the directory of type PicardHtsPath instead of String.
+     *
+     * @see #getTempFilePath(String, String, String)
+     *
+     */
+    public static PicardHtsPath getTempFilePath(final PicardHtsPath directory, String prefix, final String extension){
+        return getTempFilePath(directory.getURIString(), prefix, extension);
+    }
+
+    /**
+     * Calls getTempFilePath with the empty string as the prefix.
+     *
+     * @see #getTempFilePath(String, String, String)
+     */
     public static PicardHtsPath getTempFilePath(String directory, String extension){
         return getTempFilePath(directory, "", extension);
     }
 
+    /**
+     *
+     * Creates a temporary file on the local temporary file.
+     *
+     * @see #getTempFilePath(String, String, String)
+     */
+    public static PicardHtsPath getLocalTempFilePath(final String prefix, final String extension){
+        return getTempFilePath((String) null, prefix, extension);
+    }
+
+    /**
+     * Creates a PicardHtsPath object to a "directory" on a Google Cloud System filesystem with a randomly generated URI.
+     *
+     * Note that the notion of directories does not exist in GCS. Thus, by "directory,"
+     * we mean a path object with a randomly generated URI ending in "/", which
+     * the caller can use as a root URI/path for other files to be created e.g. via PicardHtsPath::resolve.
+     *
+     * Note that this method does *not* create an actual directory/file on GCS that one can write to, delete, or manipulate otherwise.
+     *
+     * See: https://stackoverflow.com/questions/51892343/google-gsutil-create-folder
+     *
+     * @param parentDir The root path where the new "directory" is to live e.g. "gs://hellbender-test-logs/staging/picard/test/RevertSam/".
+     * @return A PicardHtsPath object pointing to e.g. "gs://hellbender-test-logs/staging/picard/test/RevertSam/{randomly-generated-string}/"
+     */
+    public static PicardHtsPath getRandomGCSDirectory(final PicardHtsPath parentDir){
+        ValidationUtils.validateArg(parentDir.getScheme().equals(PicardBucketUtils.GOOGLE_CLOUD_STORAGE_FILESYSTEM_SCHEME), "This method is supported only for a GCS path: " + parentDir.getURIString());
+        ValidationUtils.validateArg(parentDir.getURIString().endsWith("/"), "parentDir must end in backslash '/': " + parentDir.getURIString());
+        return PicardHtsPath.fromPath(PicardBucketUtils.randomRemotePath(parentDir.getURIString(), "", "/"));
+    }
+
```

Comment:
I thought we agreed in our discussion yesterday that this method should (and I thought you said that it already did) constrain directory creation to the well-known staging bucket location that has a TTL set. As far as I can tell, it lets you create a directory anywhere.

I know the analogous method in GATK also allows you to create one anywhere as well, but that is how we accumulated the 10,000+ old temporary files floating around in GCS that date back to 2015. If we don't constrain this one, we'll replicate that same problem with Picard. Is there some reason we can't constrain this to the Picard staging location ?

Reply:
Now creates a file under a hardcoded staging location "gs://hellbender-test-logs/staging/picard/"